src/core/Form/Label/Label.tsx

```diff
@@ -40,42 +60,70 @@ const labelTextClassNames = {
 const StyledLabel = styled(
   ({
     className,
+    style,
+    contentClassName,
     theme,
     labelMode = 'visible',
-    labelSpanProps = { className: undefined },
+    wrapperProps,
     children,
     asProp = 'label',
     optionalText,
+    tooltipComponent: tooltipComponentProp,
     ...passProps
-  }: LabelProps & SuomifiThemeProp) => (
-    <HtmlDiv
-      {...(asProp ? { as: asProp } : {})}
-      className={classnames(className, baseClassName)}
-      {...passProps}
-    >
-      {labelMode === 'hidden' ? (
-        <VisuallyHidden>
-          {children}
-          {optionalText && `(${optionalText})`}
-        </VisuallyHidden>
-      ) : (
-        <HtmlSpan
-          {...labelSpanProps}
-          className={classnames(
-            labelTextClassNames.labelSpan,
-            labelSpanProps.className,
-          )}
-        >
-          {children}
-          {optionalText && (
-            <HtmlSpan className={labelTextClassNames.optionalText}>
-              {` (${optionalText})`}
+  }: LabelProps & SuomifiThemeProp) => {
+    const [wrapperRef, setWrapperRef] = useState<HTMLDivElement | null>(null);
+
+    function getTooltipComponent(
+      tooltipComponent: ReactElement | undefined,
+    ): ReactNode {
+      if (isValidElement(tooltipComponent)) {
+        return cloneElement(tooltipComponent, {
+          anchorElement: wrapperRef,
+          // trick to force tooltip to rerender every time when label changes.
+          key: Date.now(),
+        });
+      }
+      return null;
+    }
+
+    return (
+      <HtmlDivWithRef
+        className={classnames(className, baseClassName)}
+        {...wrapperProps}
+        forwardedRef={(ref: SetStateAction<HTMLDivElement | null>) =>
+          setWrapperRef(ref)
+        }
+      >
+        {labelMode === 'hidden' ? (
+          <VisuallyHidden>
```

Comment:
Do we need to pass some props to VisuallyHidden as well?

Reply:
I assume that this means, that should the tooltip's text to be read by screenreader even if the label is visually hidden.

Reply:
Not sure what exactly this need would be, but it looks like VisuallyHidden takes no props, so maybe this can be tackled later on. I'd just make sure that everything works as expected after these changes - including screen readers. Tooltip itself should not be shown imo if label mode is hidden.